### PR TITLE
Rebased: Created FloatMask to PNG writing functionality, MapImageWriter class, ImageGenerator class, and related functions

### PR DESCRIPTION
**Pull request by Penguin5**

### CODE REVIEW

File: src/java/neroxis/exporter/MapImageWriter.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            System.err.println("Error while importing the map.");
+        }
+    }
+
+    public void writeMapImages() throws IOException {
+
+        Random random = new Random();
+        FloatMask heightmapBase = map.getHeightMask(symmetrySettings);
+        heightmapBase = new FloatMask(heightmapBase, random.nextLong());
+        heightmapBase.applySymmetry(SymmetryType.SPAWN);
```
Comment (Sheikah45) [inline src/java/neroxis/exporter/MapImageWriter.java]: Why apply symmetry if you just want an image?

File: src/java/neroxis/generator/ImageGenerator.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+            base.combineBrush(new Vector2f(center + random.nextInt(variationDistance) - random.nextInt(variationDistance),
+                    center + random.nextInt(variationDistance) - random.nextInt(variationDistance)), brush2, random.nextFloat(), 1f, reducedSize);
+            base.combineBrush(new Vector2f(center + random.nextInt(variationDistance) - random.nextInt(variationDistance),
+                    center + random.nextInt(variationDistance) - random.nextInt(variationDistance)), brush3, random.nextFloat(), 1f, reducedSize);
+            base.combineBrush(new Vector2f(center + random.nextInt(variationDistance) - random.nextInt(variationDistance),
+                    center + random.nextInt(variationDistance) - random.nextInt(variationDistance)), brush4, random.nextFloat(), 1f, reducedSize);
+            base.combineBrush(new Vector2f(center + random.nextInt(variationDistance) - random.nextInt(variationDistance),
+                    center + random.nextInt(variationDistance) - random.nextInt(variationDistance)), brush5, random.nextFloat(), 1f, reducedSize);
```
Comment (Sheikah45) [inline src/java/neroxis/generator/ImageGenerator.java:162]: Uh what is the point of adding and subtracting a random?

File: src/java/neroxis/generator/ImageGenerator.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+            BinaryMask mountains = new BinaryMask(size, random.nextLong(), new SymmetrySettings(Symmetry.NONE, Symmetry.NONE, Symmetry.NONE));
+            for (int x = 0; x < 10; x++) {
+                Vector2f loc = base.getRandomPosition();
+                if(loc == null) {
+                    loc = new Vector2f(center + random.nextInt(variationDistance) - random.nextInt(variationDistance), center + random.nextInt(variationDistance) - random.nextInt(variationDistance));
```
Comment (Sheikah45) [inline src/java/neroxis/generator/ImageGenerator.java:168]: same

File: src/java/neroxis/map/BinaryMask.java [hunk 4]
```diff
@@ -503,6 +550,34 @@ public BinaryMask combineWithOffset(FloatMask other, float minValue, float maxVa
         return this;
     }
 
+    public BinaryMask combineWithOffsettoroidally(BinaryMask other, int offsetX, int offsetY, boolean centered) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/BinaryMask.java]: camelCase

File: src/java/neroxis/map/BinaryMask.java [hunk 5]
```diff
@@ -503,6 +550,34 @@ public BinaryMask combineWithOffset(FloatMask other, float minValue, float maxVa
...
+                    setYLocation += size;
+                }
+                setValueAt(setXLocation % size, setYLocation % size, other.getValueAt(x, y));
+            }
+        }
+        VisualDebugger.visualizeMask(this);
+        return this;
+    }
+
+    public BinaryMask combineWithOffsettoroidally(BinaryMask other, Vector2f loc, boolean centered) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/BinaryMask.java]: camelCase

File: src/java/neroxis/map/FloatMask.java [hunk 6]
```diff
@@ -261,6 +261,63 @@ public FloatMask add(float val) {
...
+        for (int y = 0; y < getSize(); y++) {
+            for (int x = 0; x < getSize(); x++) {
+                addValueAt(x, y, other.getValueAt(x, y) * weight);
+            }
+        }
+        VisualDebugger.visualizeMask(this);
+        return this;
+    }
+
+    public FloatMask addtoroidally(FloatMask other) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/FloatMask.java]: camelCase

File: src/java/neroxis/map/FloatMask.java [hunk 7]
```diff
@@ -261,6 +261,63 @@ public FloatMask add(float val) {
...
+        for (int y = 0; y < otherSize; y++) {
+            for (int x = 0; x < otherSize; x++) {
+                addValueAt(x % size, y % size, other.getValueAt(x, y));
+            }
+        }
+        VisualDebugger.visualizeMask(this);
+        return this;
+    }
+
+    public FloatMask addWeightedtoroidally(FloatMask other, float weight) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/FloatMask.java]: camelCase

File: src/java/neroxis/map/FloatMask.java [hunk 8]
```diff
@@ -261,6 +261,63 @@ public FloatMask add(float val) {
...
+        for (int y = 0; y < otherSize; y++) {
+            for (int x = 0; x < otherSize; x++) {
+                addValueAt(x % size, y % size, other.getValueAt(x, y) * weight);
+            }
+        }
+        VisualDebugger.visualizeMask(this);
+        return this;
+    }
+
+    public FloatMask addWithOffsettoroidally(FloatMask other, int offsetX, int offsetY, boolean centered) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/FloatMask.java]: camelCase

File: src/java/neroxis/map/FloatMask.java [hunk 9]
```diff
@@ -261,6 +261,63 @@ public FloatMask add(float val) {
...
+        for (int y = 0; y < otherSize; y++) {
+            for (int x = 0; x < otherSize; x++) {
+                addValueAt((x + offsetX + size) % size, (y + offsetY + size) % size, other.getValueAt(x, y));
+            }
+        }
+        VisualDebugger.visualizeMask(this);
+        return this;
+    }
+
+    public FloatMask addWithOffsettoroidally(FloatMask other, Vector2f loc, boolean centered) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/FloatMask.java]: camelCase

File: src/java/neroxis/map/FloatMask.java [hunk 10]
```diff
@@ -798,13 +855,53 @@ public FloatMask useBrushWithinArea(BinaryMask area, String brushName, int size,
         return this;
     }
 
+    public FloatMask useBrushWithinAreatoroidally(BinaryMask area, String brushName, int size, int numUses, float intensity) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/FloatMask.java]: camelCase

File: src/java/neroxis/map/FloatMask.java [hunk 11]
```diff
@@ -798,13 +855,53 @@ public FloatMask useBrushWithinArea(BinaryMask area, String brushName, int size,
...
+    }
+
     public FloatMask useBrushWithinAreaWithDensity(BinaryMask area, String brushName, int size, float density, float intensity) {
         int frequency = (int) (density * (float) area.getCount() / 26.21f / symmetrySettings.getSpawnSymmetry().getNumSymPoints());
         useBrushWithinArea(area, brushName, size, frequency, intensity);
         VisualDebugger.visualizeMask(this);
         return this;
     }
 
+    public FloatMask useBrushWithinAreaWithDensitytoroidally(BinaryMask area, String brushName, int size, float density, float intensity) {
```
Comment (Sheikah45) [inline src/java/neroxis/map/FloatMask.java]: camelCase

**Review by Sheikah45 — CHANGES_REQUESTED** [review]

File: src/java/neroxis/exporter/MapImageWriter.java (hunk 1 above)
Comment (Penguin5) [reply src/java/neroxis/exporter/MapImageWriter.java]: I suppose I'll remove that then.

File: src/java/neroxis/generator/ImageGenerator.java (hunk 2 above)
Comment (Penguin5) [reply src/java/neroxis/generator/ImageGenerator.java:162]: it yields a desirable spread of variation within the range of center +/- variationDistance

File: src/java/neroxis/generator/ImageGenerator.java (hunk 3 above)
Comment (Penguin5) [reply src/java/neroxis/generator/ImageGenerator.java:168]: same

File: src/java/neroxis/generator/ImageGenerator.java (hunk 2 above)
Comment (Sheikah45) [reply src/java/neroxis/generator/ImageGenerator.java:162]: You could just use random.nextInt(variationDistance * 2) - variationDistance

Comment (Sheikah45) [reply src/java/neroxis/generator/ImageGenerator.java:162]: also random.nextGaussian

Comment (Sheikah45) [reply src/java/neroxis/generator/ImageGenerator.java:162]: but depends on if you want the uniform distribution or not

Comment (Penguin5) [reply src/java/neroxis/generator/ImageGenerator.java:162]: I don't want uniform distrubution

**Review by Sheikah45 — APPROVED** [review]

**Pull request merged**